### DISCUSSION TITLE
mutiple views using paraview

### DISCUSSION THREAD

**Discussion by AliE89** [Kitware/trame#374] [Q&A] [answered]
Hi Jourdain,

Sorry it's me again.  I am trying to visualise two different VTM files in two separate windows using paraview and trame. 
I took advantage of example 06_vtk/03_multiviews and I managed to create a row with 2 different columns. So far so good. For each column, I am assigning different view or at least this is what I am thinking. Unfortunately, it looks like the same view is passed through both windows.  Here what I am doing:
-	Importing vtm1
-	Importing vtm2
-	Hiding vtm2 and assign the view to window1
-	Hiding vtm1 and assign view to window2
The code looks like this:
```
	For vtm in vtms:
	       …
                      with vuetify.VCol(classes="pa-0 ma-1"):
                            print(idx, paraview_state.active_filter)
                            HideAll()
                            vtm.show()
                            view = vtm.render_view
                            view.MakeRenderWindowInteractor(True)
                            Render(view)
                            html_view = paraview.VtkRemoteView(view)
                            html_views.append(html_view)
                            ctrl.update_views.add(html_view.update)
                            ctrl.reset_camera.add(html_view.reset_camera)

```
That vtm.show() is a little wrapper that take a target filter in the pipeline and hide it.
I must be doing something odd here. Is there any special trick to perform this operation in your opinion?
Many thanks! 😊

**Comment by jourdain** [comment]
I need to make it automatic when not provided as I do for other components, but for now, you need to assign a unique name to each of your views using the `ref="..."` attribute.

**Reply by jourdain** [reply]
FYI, the ref arg should not be needed anymore if you use [that release](https://github.com/Kitware/trame-vtk/releases/tag/v2.6.2)

**Comment by AliE89** [comment]
Good mornig,

Do you mean here?paraview.VtkRemoteView(view, ref=view_1

Cheers

Il Gio 16 Nov 2023, 03:37 Sebastien Jourdain (Kitware) <
***@***.***> ha scritto:

**Reply by AliE89** [reply, marked as answer]
It is working. However, I must create two seprate views with CreateView('RenderView') and then pass it to VtkRemoteView, giving a name. 

I'll keep playing with this as likely I'll hit something else :)

**Reply by AliE89** [reply]
It was a good day of testing, I manage to do a lot of stuff. However, all the representations work but the LIC. Does this sound a bell?

**Reply by jourdain** [reply]
It should but you may have to load the plugin to have it available

**Reply by AliE89** [reply]
I tried to laod the plugin (paraview 5.12 says that now it's built in), but no luck. Since other representation works OK, I am not sure it's an error on my side (but probably it is -.-)

**Reply by jourdain** [reply]
Did you use the trace to get the Python code to use?

**Reply by AliE89** [reply]
Sure thing, here my code:

```
display = GetDisplayProperties(active_filter, render_view)
display.SetRepresentationType('Surface LIC')
```
That should trigger the LIC.  Any other representation (Surface, Wireframe, Surface With Edges) works nicely.

Cheers

**Reply by jourdain** [reply]
Just tested and it is working for me...

```python
# export PYTHONPATH=/Applications/ParaView-5.12.0-RC1.app/Contents/Python/
# export DYLD_LIBRARY_PATH=/Applications/ParaView-5.12.0-RC1.app/Contents/Libraries/

from paraview import simple
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import paraview, html

# create a new 'IOSS Reader'
disk_out_refex2 = simple.IOSSReader(registrationName='disk_out_ref.ex2', FileName=['/Applications/ParaView-5.12.0-RC1.app/Contents/examples/disk_out_ref.ex2'])

# get active view
renderView1 = simple.GetActiveViewOrCreate('RenderView')

# show data in view
disk_out_refex2Display = simple.Show(disk_out_refex2, renderView1, 'UnstructuredGridRepresentation')

# trace defaults for the display properties.
disk_out_refex2Display.Representation = 'Surface'


# update the view to ensure updated data information
renderView1.Update()

# change representation type
disk_out_refex2Display.SetRepresentationType('Surface LIC')


server = get_server(client_type="vue3")
with DivLayout(server) as layout:
    layout.root.style = "width: 100vw; height: 100vh; margin: 0; padding: 0;"
    reset_camera = paraview.VtkRemoteView(renderView1).reset_camera
    html.Button("Reset", style="position: absolute; top: 10px; left: 10px;", click=reset_camera)

server.start()


```

**Reply by AliE89** [reply]
Thanks a lot for this. 

I have tried your code, but I cannot get anything as mine. I should mention that I am on 2 render windows. Other than that, the only difference is I am using vue2 and not veu3. :/

I'll keep digging.

**Reply by jourdain** [reply]
You can change my code to use "vue2" as it has no dependency to a specific version of vue.

Either way, if "my code works" on your side (after fixing file paths), that means the capability is supported and should just work on your end.

**Reply by AliE89** [reply]
Hi Sèbastiene,

Sorry to get back you late on this, but I have spent some time in doing some more tests on my side. 

I tried your example and indeed it worked ok on my system. 

I start thinking there is something odd happening with *.vtm format. As soon has I jsut change the import file using your exact same pipeline, I am not able to se LIC anymore.

**Reply by AliE89** [reply]
Just to be more helpful, I attach here my silly sample code

```
from paraview.web import venv
from paraview.simple import *
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import paraview, html


vtm_file = XMLMultiBlockDataReader(registrationName='test', FileName=['test.vtm'])
vtm_file.PointArrayStatus = ['test_var']

renderView1 = GetActiveViewOrCreate('RenderView')
disk_out_refex2Display = Show(vtm_file, renderView1)
disk_out_refex2Display.Representation = 'Surface'
disk_out_refex2Display.SetRepresentationType('Surface LIC')
renderView1.Update()


server = get_server(client_type="vue3")
with DivLayout(server) as layout:
    layout.root.style = "width: 100vw; height: 100vh; margin: 0; padding: 0;"
    reset_camera = paraview.VtkRemoteView(renderView1).reset_camera
    html.Button("Reset", style="position: absolute; top: 10px; left: 10px;", click=reset_camera)

server.start()
```

**Reply by AliE89** [reply]
For this particular case, I guess I have found the issue:

`disk_out_refex2Display.SelectInputVectors = ['POINTS', 'test_var']`

I needed to add this line to trigger the input vector. Let me try with his big brother.

**Reply by jourdain** [reply]
So I guess it is now working for you.

**Comment by AliE89** [comment]
Yes. All good.  The error is 99% of the time in front of the screen!

Il Gio 30 Nov 2023, 17:08 Sebastien Jourdain (Kitware) <
***@***.***> ha scritto:
